lib/xapi-stdext-unix/unixext.ml

```diff
@@ -161,12 +161,17 @@ let string_of_file file_path = Buffer.contents (buffer_of_file file_path)
 let atomic_write_to_file fname perms f =
   let module Filenameext = Xapi_stdext_std.Filenameext in
   let tmp = Filenameext.temp_file_in_dir fname in
-  Unix.chmod tmp perms;
+  let tmp_perms = 0o600 in
+  Unix.chmod tmp tmp_perms;
   finally
     (fun () ->
-       let fd = Unix.openfile tmp [Unix.O_WRONLY; Unix.O_CREAT] perms (* ignored since the file exists *) in
+       (* file permissions in openfile are ignored since the file exists *)
+       let fd = Unix.openfile tmp [Unix.O_WRONLY; Unix.O_CREAT] tmp_perms in
        let result = finally
-           (fun () -> f fd)
+           (fun () ->
+             let res = f fd in
+             Unix.fchmod fd perms;
```

Comment:
There are more things we can do to improve the reliability of this atomic rename, i.e. if we also want to make it durable not just atomic.

As it is currently written you would likely end up with some 0-sized files on XFS after a power failure/crash.

If you want to ensure that you have either the old data, or the new data, and nothing inbetween then I think the most reliable way that is filesystem independent:
* write new file to a tmpfile as is done here
* fsync the tmpfile
* close and rename
* open the parent directory readonly
* fsync the parent directory

I'll try to find an article that described this (I think it was on LWN.net, or a blogpost from the ext4 author)

Reply:
See https://thunk.org/tytso/blog/2009/03/15/dont-fear-the-fsync/ which recommends write/fsync/close/rename for atomicity, and an optional fsync of parent dir if you also want durability, i.e. to guarantee that after a crash the new file would be present (even though you fsynced the tmp file, after you do the rename and you crash you still don't know whether you get the old file or the new file). In general it may not matter whether you get the old or new file (as long as you don't get a corrupt or 0-sized file inbetween), but if you then do other (durable) operations based on the outcome of the first, then having durability guarantees from the 1st operation would be useful.

Of course fsync comes with a performance cost, especially on ext3 which we use, but thats the cost of not loosing data.

Reply:
https://www.slideshare.net/nan1nan1/eat-my-data (also some older slides that explain the problem about close/rename well, coming from a MySQL developer).